applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+
+	fetch_event = new_config_fetch_event(sizeof(struct params_chmap));
+	__ASSERT_NO_MSG(fetch_event != NULL);
+	qos_chmap = (struct params_chmap *) fetch_event->dyndata.data;
+
+	k_mutex_lock(&m_data_access_mutex, K_FOREVER);
+	memcpy(qos_chmap->chmap, m_current_chmap, sizeof(m_current_chmap));
+	k_mutex_unlock(&m_data_access_mutex);
+
+	return fetch_event;
+}
+
+static struct config_fetch_event *fetch_event_qos_blacklist(void)
+{
+	struct config_fetch_event *fetch_event;
+	struct params_blacklist *qos_blacklist;
+
+	fetch_event = new_config_fetch_event(sizeof(struct params_blacklist));
+	__ASSERT_NO_MSG(fetch_event != NULL);
+	qos_blacklist = (struct params_blacklist *) fetch_event->dyndata.data;
+	memset(qos_blacklist, 0, sizeof(struct params_blacklist));
+
+	qos_blacklist->wifi_chn_bitmask = chmap_filter_wifi_blacklist_get();
+
+	return fetch_event;
+}
+
+static void update_parameters(
+	struct params_ble *p_qos_ble_params,
+	struct params_wifi *p_qos_wifi_params)
+{
+	struct chmap_filter_params *p_params;
+
+	p_params = k_malloc(sizeof(struct chmap_filter_params));
+	if (p_params == NULL) {
+		LOG_ERR("OOM error");
+		return;
+	}
+
+	chmap_filter_params_get(mp_chmap_inst, p_params);
+
+	if (p_qos_ble_params != NULL) {
+		p_params->min_channel_count =
+			p_qos_ble_params->min_channel_count;
+		p_params->ble_weight_crc_ok =
+			p_qos_ble_params->weight_crc_ok;
+		p_params->ble_weight_crc_error =
+			p_qos_ble_params->weight_crc_error;
+		p_params->ble_block_threshold =
+			p_qos_ble_params->ble_block_threshold;
+		p_params->eval_max_count =
+			p_qos_ble_params->eval_max_count;
+		p_params->eval_duration =
+			p_qos_ble_params->eval_duration;
+		p_params->maintenance_sample_count =
+			p_qos_ble_params->sample_count_min;
+		p_params->eval_keepout_duration =
+			p_qos_ble_params->eval_keepout_duration;
+		p_params->eval_success_threshold =
+			p_qos_ble_params->eval_success_threshold;
+	}
+	if (p_qos_wifi_params != NULL) {
+		p_params->wifi_rating_inc =
+			p_qos_wifi_params->wifi_rating_inc;
+		p_params->wifi_present_threshold =
+			p_qos_wifi_params->wifi_present_threshold;
+		p_params->wifi_active_threshold =
+			p_qos_wifi_params->wifi_active_threshold;
+	}
+
+	k_mutex_lock(&m_data_access_mutex, K_FOREVER);
+	if (m_new_params != NULL) {
+		k_free(m_new_params);
+	}
+	m_new_params = p_params;
+	k_mutex_unlock(&m_data_access_mutex);
+}
+
+static void update_blacklist(struct params_blacklist *p_blacklist)
+{
+	atomic_set(&m_new_blacklist, p_blacklist->wifi_chn_bitmask);
+}
+
+static int hci_chmap_cmd_send(const u8_t *p_chmap)
+{
+	struct bt_hci_cp_le_set_host_chan_classif *cp;
+	struct net_buf *buf;
+
+	__ASSERT_NO_MSG(p_chmap != NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+				sizeof(*cp));
+	if (!buf) {
+		return -1;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	if (!cp) {
+		return -1;
+	}
+
+	memcpy(cp->ch_map, p_chmap, sizeof(cp->ch_map));
+	__ASSERT_NO_MSG(sizeof(cp->ch_map) == 5);
```

Comment:
Assuming this is protecting from stack overflow, isn't it a bit late to check the size after the memcpy?

Reply:
There's no risk of overflowing the dest buffer in the memcpy, but the assert is not terribly useful either. The assert was intended as a sanity check during development (expected BLE channel map size is always 5). Would make more sense to instead have a static assert to validate the 5-byte sizes.